### ISSUE TITLE
Do not require fragments to have `id` in their root in `require-id-when-available`

### ISSUE THREAD

**Issue by panrafal** [dotansimha/graphql-eslint#731]
Fragments should not be required to include `id` field on their root selection set, as this is the query's responsibility.

This causes a couple of issues:
- Fragments have to include `id` which they may not need at all
- If multiple fragments are inlined, all of them need to include the `id`
- If `id` is missing on the query and all used fragments, errors are reported for the query and all fragments, making the problem much worse than it is

Consider this example:

```
fragment UserAvatar {
  firstName
  lastName
  avatar
}
fragment UserManager {
  manager {
    id
    ...UserAvatar
  }
}

query GetCurrentUser {
  me {
    id
    ...UserAvatar
    ...UserManager
  }
}
```

Both fragments define data requirements of simple UI components - they don't need the User's `id` to work at all.
This should be added on the query level, as this is the cache's requirement.

An exception to this is when deeper objects are queried in fragments. Queries don't know about the fragment internals,
so in this case, it's fragment's responsiblity to query the `id`.

I've made a working change on my fork as we're using this in a big project: https://github.com/dotansimha/graphql-eslint/compare/master...panrafal:panrafal/dont-require-id-on-fragment-root

**Comment by B2o5T** [comment]
Hi @panrafal, thanks for the report, this will be tricky to implement, but I'll work on this to make `graphql-eslint` better 🥳

**Comment by panrafal** [comment]
Hey @B2o5T , maybe I'm not seeing something, but the change is actually very simple and I have a [PR ready](https://github.com/dotansimha/graphql-eslint/compare/master...panrafal:panrafal/dont-require-id-on-fragment-root). 
It would be backwards-compatible too, as it relaxes the rule a bit, while still making sure that all queried objects have their IDs.

**Comment by B2o5T** [comment]
Hey @panrafal the `tricky` part is that we can make this rule smarter without ignoring `id` field for fragments by checking if the fragment is imported by some operation and at least 1 operation requires his `id` field

**Comment by dotansimha** [comment]
> Hey @panrafal the `tricky` part is that we can make this rule smarter without ignoring `id` field for fragments by checking if the fragment is imported by some operation and at least 1 operation requires his `id` field.
> 
> @dotansimha what do you think?

Agree, I think we can consider adding support for parserServices - to allow us to check if a fragment is used and get a list of operations that are using it and then validate it there. 
It might add some complexity because fragments might be used by other fragments that are used in a query - so this needs some way to "flatten" the query that uses each fragment, and then check if the `id` is used there?

**Comment by dobesv** [comment]
I have run into this issue and I would propose this rule should work as follows:

1. Only report this error on queries and mutations, not fragments
2. Report any "path" in the query that could have had `_id` but didn't after taking fragment spreads into account

Trying to apply the rule to fragments is confusing because the fragment doesn't tell the whole story - likely a fragment without `_id` is intended to be used in a query that does have `_id` listed directly or in another fragment.

**Comment by dotansimha** [comment]
Fixed in `@graphql-eslint/eslint-plugin@3.5.0`.
